File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(heights) {
    let maxArea = 0;
    
    let [left, right] = [0, heights.length - 1];
    
    while (left < right) {
        tempArea = (right - left) * Math.min(heights[left], heights[right]);
        if (tempArea > maxArea) maxArea = tempArea;
        
        if (heights[left] < heights[right]) {
            left++;
        } else if (heights[left] > heights[right]) {
            right--;
        } else {
            if (heights[left + 1] < heights[right - 1]) {
                right--;
            } else {
                left++;
            }
        }
    }
    
    return maxArea;
};

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, heights: list[int]) -> int:
        max_water = 0

        left, right = 0, len(heights) - 1

        while left < right:
            left_height = heights[left]
            right_height = heights[right]
            curr_water = (right - left) * min(left_height, right_height)
            max_water = max(max_water, curr_water)

            if left_height < right_height:
                left += 1
            else:
                right -= 1

        return max_water
